### ISSUE TITLE
Allow Option params in systems

### ISSUE THREAD

**Issue by TsujimotoYukasuke** [bevyengine/bevy#7516]
## What problem does this solve or what need does it fill?

I often create functions that get called once as startup systems that I later on may want to reuse somewhere else in my code. For example, a `spawn_weapon` function that I might want to use at the start of the game, but obviously may end up wanting to use later in the game as well.

## What solution would you like?

I want to have the option to place an Option type within a system. By default, the Bevy ECS would place a "None" value into the parameter, allowing me to have the option to provide default functionality.

Later on I can use that function but place "Some" instead to have custom functionality.

So, for example:

```rust
use bevy::prelude::*;

// ...

fn spawn_weapon(
    mut commands: Commands,
    rooms: Res<Rooms>,
    spawn_location: Option<Vec3>,
) {
    let transform = match spawn_location {
        Some(location) => transform.from_translation(location),
        None => rooms.current().default_weapon_spawn_location(),
    }

    commands.spawn(WeaponBundle {
        transform,
        ..Default::default()
    });
}
```

## What alternative(s) have you considered?

The most straightforward way to do this currently is to create a utility function (`spawn_weapon`) and then the actual startup system (`spawn_weapon_on_startup`).

**Comment by alice-i-cecile** [comment]
Would a Local<Option<T>> work?

**Comment by james7132** [comment]
Alternatively if you need external input `In<Option<T>>` works as well.

**Comment by TsujimotoYukasuke** [comment]
Both of those options may work, though I think there's some significant semantic differences between the three.

I think that having a type that clearly displays itself as an optional parameter that the Bevy ecs will default on would be a nice addition.

With a Local I feel the intent is more aligned with having an optional value that will be modified by the system itself over a series of calls. In would be more appropriate for an optional type that gets passed by another system.

**Comment by alice-i-cecile** [comment]
Hmm: it feels like https://github.com/bevyengine/bevy/pull/7162 + an `if let` or a `?` to handle the empty case is the right tool for this then. I'm still struggling to understand the exact pattern you want here though.

**Comment by JoJoJet** [comment]
@TsujimotoYukasuke can you elaborate on how your proposed `Option<>` params might work? In the example you gave, it seems like it will always just be `None`. When/how would the default value get replaced with `Some`?

**Comment by TsujimotoYukasuke** [comment]
Yes, you are correct, it's intended for the default to always be `None`.

The only time `Some` would be passed is if the function *is not* being called by ecs (i.e. calling it manually).

**Comment by JoJoJet** [comment]
Ah, I see. If you need to configure a system, our preferred pattern for that is to return a system from a function.

```rust
// This function is *not* a system -- it *returns* a system when called.
fn spawn_weapon(
    spawn_location: Option<Vec3>,
) -> impl FnMut(Commands, Res<Rooms>) {
    |mut commands, rooms| {
        let transform = match spawn_location {
            Some(location) => transform.from_translation(location),
            None => rooms.current().default_weapon_spawn_location(),
        }

        commands.spawn(WeaponBundle {
            transform,
            ..Default::default()
        });
    }
}

// Use the default spawn location.
app.add_system(spawn_weapon(None));

// Manually call the function from within another system.
fn some_other_system(commands: Commands, rooms: Res<Rooms>) {
    // Configure the `spawn_weapon` system to use a specific spawn location.
    let spawn_weapon_at = spawn_weapon(Some(Vec3::ONE));
    spawn_weapon_at(commands, rooms);
}
```

Do you think this would work in your case?

**Comment by TsujimotoYukasuke** [comment]
Definitely would, and also really cool.

I guess my suggestion isn't really needed in that case, though I do think that if we can find a way to make that pattern a bit less verbose (for example, through a macro or something) it'd be great for those newer to Rust and Bevy.

**Comment by alice-i-cecile** [comment]
Docs would also help a lot: this is a cool pattern, but hard to discover.

**Comment by TsujimotoYukasuke** [comment]
Can `impl FnOnce` be converted into a system? If not, then I think I've found an issue with that pattern.

I'm currently making yet another snake clone, which is what made me create this issue in the first place. Just tried using the pattern, but had to use `impl FnOnce` instead of `impl FnMut` since I wanted to move one of my variables into the closure. Here's the code:

```rust
// ...

pub struct SnakePlugin;
impl Plugin for SnakePlugin {
    fn build(&self, app: &mut App) {
        app.add_fixed_timestep(Duration::from_millis(START_MOVEMENT_STEP), "movement_step")
            .add_startup_system(spawn_snake(None, None)) // <--- Use of the pattern here
            .add_fixed_timestep_system("movement_step", 0, move_snakes)
            .add_system_to_stage(CoreStage::PostUpdate, update_segments);
    }
}

// ...

fn spawn_snake(position: Option<GridPosition>, snake: Option<Snake>) -> impl FnOnce(Commands) {
    move |mut commands| {
        let position = match position {
            Some(position) => position,
            None => GridPosition::default(),
        };

        let mut snake = match snake { // <--- If we use FnMut then this will throw an error due to it being a borrowed value
            Some(snake) => snake,
            None => Snake::default(),
        };

        let snake_head = commands
            .spawn(SpriteBundle {
                sprite: Sprite {
                    color: snake.color,
                    ..default()
                },
                ..default()
            })
            .insert(SnakeSegment)
            .insert(GridPosition::default())
            .id();

        snake.segments.push((snake_head, position));
        commands.spawn(snake);
    }
}
```

My first thought is to use `.to_owned` or `.clone`, but from what I understand, that's a bit of an anti-pattern.

**Comment by JoJoJet** [comment]
> My first thought is to use `.to_owned` or `.clone`, but from what I understand, that's a bit of an anti-pattern.

Not really, no. In performance-critical code it can be good to avoid clones when possible, but in most cases cloning isn't really worth worrying about.

What does your `Snake` struct look like? You might be able to just `#[derive(Copy)]` it and not have to worry about ownership at all.

**Comment by TsujimotoYukasuke** [comment]
I believe you're right, it's composed of a bunch of structs that are cloned anyway. But I also feel that there may be a situation where someone might want to use the pattern in this way anyway.

I have no idea how trivial it is to implement this pattern for `FnOnce`, but if it happens to be a trivial solution I think it would be a good addition.

**Comment by JoJoJet** [comment]
> I have no idea how trivial it is to implement this pattern for `FnOnce`, but if it happens to be a trivial solution I think it would be a good addition.

Systems need to be called multiple times (once each frame), so in general it's not possible to have a `FnOnce` system. I suppose it's *theoretically* possible to have a `FnOnce` startup system (since startup systems only get called once) -- however this would be a really high-impact change and probably not worth all that much since it's pretty easy to work around this issue.

**Comment by TsujimotoYukasuke** [comment]
Fair enough

**Comment by JoJoJet** [comment]
> Docs would also help a lot: this is a cool pattern, but hard to discover.

This is currently documented in [`Local`](https://docs.rs/bevy/latest/bevy/prelude/struct.Local.html). I suppose it could be in a more visible place, but I'm not sure where. The cheatbook, perhaps?
